src/tensors/cpu/tensor_operators.cpp

```diff
@@ -696,19 +719,50 @@ void Select(Tensor out,
 
   matchOrAbort<IndexType>(indices->type());
 
-  // @TODO: make this efficient
   functional::Shape outShape = out->shape();
   functional::Shape inShape  = in->shape();
   functional::Shape idxShape = indices->shape();
-  int length = outShape.elements();
 
-  functional::Array<int, functional::Shape::size()> dims;
   int axisCPU = (int)(axis + functional::Shape::size() - out->shape().size());
 
-#if 0 // buggy but not really used?
-  if(axisCPU == 2 && outShape == idxShape) // specialization for axis==2 when there is no broadcasting, @TODO to be removed once we have a faster implementation below
-    return SelectAxis2(out, in, indices);
-#endif
+  // Are all index dimensions 1 after the axis?
+  bool flatIndices = true;
+  // Total dimensionality of input and output after the axis.
+  int afterAxis = 1;
+  for (int i = axisCPU + 1; i < functional::Shape::size(); ++i) {
+    afterAxis *= outShape[i];
+    if (idxShape[i] != 1) {
+      flatIndices = false;
+    }
+  }
+  /* Faster version based on copying. Requirements:
+   * input is contiguous for every dimension after the axis.
+   * output is contiguous for every dimension after the axis.
+   * indices have shape 1 for every dimension after the axis.
+   */
+  if (afterAxis == inShape.stride(axisCPU) && afterAxis == outShape.stride(axisCPU) && flatIndices) {
+    const int end = outShape.dim(axisCPU);
+    const int outStride = outShape.stride(axisCPU);
+    const int idxStride = idxShape.bstride(axisCPU);
+    // Loop over all dimensions before the axis.
+    LoopBeforeAxis(outShape, inShape, idxShape, axisCPU,
+        [out, in, indices, afterAxis, end, outStride, idxStride](int outBase, int inBase, int idxBase) {
+          // Loop over the axis dimension.
+          for (int i = 0; i < end; ++i) {
+            int index = indices->data<IndexType>()[idxBase];
+            // Loop over all dimensions after the axis.
+            std::copy(in->data() + inBase + index * afterAxis, in->data() + inBase + index * afterAxis + afterAxis, out->data() + outBase);
+            outBase += outStride;
+            idxBase += idxStride;
+          }
+        });
+    return;
```

Comment:
Let's do an `if {} else {}` here instead of the empty return.  Could you also mention that the code below is execute if the above conditions are not met and that this code is generic but slow.

Reply:
Maybe let's also add a positive and a negative example where each branch is used? There are a couple of them in tests_operators, I think at least one will still go to the default version?